libnavigation-base/src/main/java/com/mapbox/navigation/base/trip/model/EHorizon.kt

```diff
@@ -0,0 +1,108 @@
+package com.mapbox.navigation.base.trip.model
+
+/**
+ * The EHorizon
+ *
+ * @param start
+ */
+class EHorizon private constructor(
+    val start: Edge
+) {
+
+    /**
+     * Get the current edge given the position
+     */
+    fun current(position: EHorizonPosition): Edge {
+        var edge: Edge? = start
+        while (edge?.id != position.edgeId) {
+            edge = edge?.out?.find { e -> e.level == 0.toByte() }
```

Comment:
I guess this line is very tricky. It's really easy to get into infinite loop here
Sample code to demonstrate the problem
```
class Edge(
    val id: Int,
    val level: Int,
    val out: List<Edge>?
)

// a graph like 
// 1 - 2
// \               
// 3
// will break this function
fun current(start: Edge, edgeId: Int): Edge {
    var edge: Edge? = start
    while (edge?.id != edgeId) {
        edge = edge?.out?.find { e -> e.level == 0 }
    }
    return edge
}

fun main() {
    val thirdEdge = Edge(2, 0, null)
    val secondEdge = Edge(1, 0, null) // no matter how deep is the structure. If it doesn't have a connection to the third edge finding current edge won't succeed.
    val firstEdge = Edge(0, 0, listOf(secondEdge, thirdEdge))
    val currentEdge = current(firstEdge, 2)
    println(currentEdge)
}
```

Reply:
A possible solution
```
fun current(start: Edge, edgeId: Int): Edge {
    val visitedEdges = LinkedList<Edge>()
    visitedEdges.push(start)
    while (visitedEdges.isNotEmpty()) {
        val current = visitedEdges.poll()
        if (current.id != edgeId) {
            val mppEdges = current?.out?.filter { e -> e.level == 0 }
            mppEdges?.let { visitedEdges.addAll(it) }
        } else {
            return current
        }
    }
    throw IllegalArgumentException("No edge with id $edgeId")
}
```

Reply:
Nice catch! Setting up the project to have some unit tests https://github.com/mapbox/mapbox-navigation-android/pull/3777

Should capture any issue we have with a test, and then fix it